File: tools/utils/cli-args.rb
require 'optparse'

module CLIArgs
  VALID_MODES = %w[dev test stage production]

  def self.parse(argv)
    options = { mode: nil, init_git_for_patch: false, production_mozbuild: nil, help: false }
    OptionParser.new do |opts|
      opts.on('--mode=MODE', String) { |m| options[:mode] = m }
      opts.on('--production-mozbuild=PHASE', String) { |p| options[:production_mozbuild] = p }
      opts.on('--init-git-for-patch') { options[:init_git_for_patch] = true }
      opts.on('-h', '--help') { options[:help] = true }
    end.parse!(argv)
    options
  end

  def self.validate(options)
    return if options[:help]
    if options[:production_mozbuild] && options[:mode] && options[:mode] != 'production'
      raise '`--production-mozbuild` is not compatible with `--mode=`.'
    end
  end

  def self.map_to_build_options(options)
    mode = VALID_MODES.include?(options[:mode]) ? options[:mode] : 'dev'
    phase = case options[:production_mozbuild]
            when 'before' then 'before'
            when 'after' then 'after'
            end
    mode = 'production' if phase
    {
      mode: mode,
      init_git_for_patch: options[:init_git_for_patch],
      production_mozbuild_phase: phase,
      launch_browser: %w[dev test stage].include?(mode)
    }
  end
end

File: tools/utils/cli-args.ts
// tools/utils/cli-args.ts
/**
 * CLI argument parsing, validation, and mapping for Noraneko Build System.
 */

import { parseArgs } from "@std/cli/parse-args";
import type { BuildOptions, ProductionMozbuildPhase } from "./types.ts";
import type { BuildMode } from "./types.ts";

/**
 * Raw CLI arguments structure.
 */
export interface RawCLIArgs {
  mode?: string;
  "init-git-for-patch"?: boolean;
  "production-mozbuild"?: string;
  help?: boolean;
  h?: boolean;
  [key: string]: unknown;
}

/**
 * Parse CLI arguments into a structured object.
 */
export function parseCLIArgs(args: string[]): RawCLIArgs {
  return parseArgs(args) as RawCLIArgs;
}

/**
 * Validate and normalize CLI arguments.
 * Throws on invalid combinations.
 */
export function validateCLIArgs(args: RawCLIArgs): void {
  if (args.help || args.h) {
    // Help is handled by CLI.showHelp
    return;
  }
  if (args["production-mozbuild"] && args.mode && args.mode !== "production") {
    throw new Error(
      "`--production-mozbuild` is not compatible with `--mode=`.",
    );
  }
}

/**
 * Map CLI arguments to BuildOptions.
 */
export function mapCLIArgsToBuildOptions(args: RawCLIArgs): BuildOptions {
  // Only allow valid BuildMode values
  const validModes: BuildMode[] = ["dev", "test", "stage", "production"];
  let mode: BuildMode = validModes.includes(args.mode as BuildMode)
    ? (args.mode as BuildMode)
    : "dev";
  let productionMozbuildPhase: ProductionMozbuildPhase | undefined;

  if (args["production-mozbuild"] === "before") {
    mode = "production";
    productionMozbuildPhase = "before";
  } else if (args["production-mozbuild"] === "after") {
    mode = "production";
    productionMozbuildPhase = "after";
  }

  const launchBrowser = mode === "dev" || mode === "test" || mode === "stage";

  return {
    mode,
    initGitForPatch: !!args["init-git-for-patch"],
    productionMozbuildPhase,
    launchBrowser,
  };
}
